File: mahjong_utils/yaku/yakuman.py
from typing import Set

from .extra import tenhou, chihou
from .yaku import Yaku

kokushi = Yaku("kokushi", 13, 13, True)
suanko = Yaku("suanko", 13, 13, True)
daisangen = Yaku("daisangen", 13, 0, True)
tsuiso = Yaku("tsuiso", 13, 0, True)
shousushi = Yaku("shousushi", 13, 0, True)
lyuiso = Yaku("lyuiso", 13, 0, True)
chinroto = Yaku("chinroto", 13, 0, True)
sukantsu = Yaku("sukantsu", 13, 0, True)
churen = Yaku("churen", 13, 13, True)
daisushi = Yaku("daisushi", 26, 0, True)
churen_nine_waiting = Yaku("churen_nine_waiting", 26, 26, True)
suanko_tanki = Yaku("suanko_tanki", 26, 26, True)
kokushi_thirteen_waiting = Yaku("kokushi_thirteen_waiting", 26, 26, True)

all_yakuman: Set[Yaku] = {
    tenhou, chihou,
    kokushi, suanko, daisangen, tsuiso, shousushi, lyuiso, chinroto, sukantsu, churen,
    daisushi, churen_nine_waiting, suanko_tanki, kokushi_thirteen_waiting
}

__all__ = (
    "all_yakuman",
    "tenhou", "chihou",
    "kokushi", "suanko", "daisangen", "tsuiso", "shousushi", "lyuiso", "chinroto", "sukantsu", "churen",
    "daisushi", "churen_nine_waiting", "suanko_tanki", "kokushi_thirteen_waiting"
)

File: src/commonMain/kotlin/mahjongutils/yaku/Yakuman.kt
package mahjongutils.yaku

import mahjongutils.models.Tile
import mahjongutils.models.TileType
import mahjongutils.models.hand.KokushiHoraHandPattern

val Kokushi = Yaku("Kokushi", 13, 13, true) { pattern ->
    pattern is KokushiHoraHandPattern && !pattern.thirteenWaiting
}

val Suanko = Yaku("Suanko", 13, 13, true, ankoSeriesCheckerFactory(4, tanki = false))

val Daisangen = Yaku("Daisangen", 13, 0, true, sangenSeriesCheckerFactory(3, sangenJyantou = false))

val Tsuiso = Yaku("Tsuiso", 13, 0, true) { pattern ->
    pattern.tiles.all { it.type == TileType.Z }
}

val Shousushi = Yaku("Shousushi", 13, 0, true, sushiSeriesCheckerFactory(3, windJyantou = true))

private val lyuTiles = Tile.parseTiles("23468s6z").toSet()

val Lyuiso = Yaku("Lyuiso", 13, 0, true) { pattern ->
    pattern.tiles.all { it in lyuTiles }
}

val Chinroto = Yaku("Chinroto", 13, 0, true, yaochuSeriesCheckerFactory(shuntsu = false, z = false))

val Sukantsu = Yaku("Sukantsu", 13, 0, true, kantsuSeriesCheckerFactory(4))

val Churen = Yaku("Churen", 13, 13, true, churenSeriesCheckerFactory(nineWaiting = false))

val Daisushi = Yaku("Daisushi", 26, 0, true, sushiSeriesCheckerFactory(4, windJyantou = false))

val ChurenNineWaiting = Yaku("ChurenNineWaiting", 26, 26, true, churenSeriesCheckerFactory(nineWaiting = true))

val SuankoTanki = Yaku("SuankoTanki", 26, 26, true, ankoSeriesCheckerFactory(4, true))

val KokushiThirteenWaiting = Yaku("KokushiThirteenWaiting", 26, 26, true) { pattern ->
    pattern is KokushiHoraHandPattern && pattern.thirteenWaiting
}
